sale_start_end_dates/models/sale.py

```diff
@@ -0,0 +1,141 @@
+        if (self.default_start_date and self.default_end_date and
+                self.default_start_date > self.default_end_date):
+            self.default_start_date = self.default_end_date
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    start_date = fields.Date(
+        string='Start Date', readonly=True,
+        states={'draft': [('readonly', False)]})
+    end_date = fields.Date(
+        string='End Date', readonly=True,
+        states={'draft': [('readonly', False)]})
+    number_of_days = fields.Integer(string='Number of Days')
+    must_have_dates = fields.Boolean(
+        related='product_id.must_have_dates',
+        readonly=True
+    )
+
+    @api.constrains('start_date', 'end_date', 'number_of_days')
+    def _check_start_end_dates(self):
+        if self.must_have_dates:
+            if not self.end_date:
+                raise ValidationError(_(
+                    "Missing End Date for sale order line with "
+                    "Product '%s'.") % (self.product_id.name))
+            if not self.start_date:
+                raise ValidationError(_(
+                    "Missing Start Date for sale order line with "
+                    "Product '%s'.") % (self.product_id.name))
+            if not self.number_of_days:
+                raise ValidationError(_(
+                    "Missing number of days for sale order line with "
+                    "Product '%s'.") % (self.product_id.name))
+            if self.start_date > self.end_date:
+                raise ValidationError(_(
+                    "Start Date should be before or be the same as "
+                    "End Date for sale order line with Product '%s'.")
+                    % (self.product_id.name))
+            if self.number_of_days < 0:
+                raise ValidationError(_(
+                    "On sale order line with Product '%s', the "
+                    "number of days is negative ; this is not allowed.")
+                    % (self.product_id.name))
+            days_delta = (self.end_date - self.start_date).days + 1
+            if self.number_of_days != days_delta:
+                raise ValidationError(_(
+                    "On the sale order line with Product '%s', "
+                    "there are %d days between the Start Date (%s) and "
+                    "the End Date (%s), but the number of days field "
+                    "has a value of %d days.")
+                    % (self.product_id.name, days_delta, self.start_date,
+                        self.end_date, self.number_of_days))
+
+    @api.multi
+    def _prepare_invoice_line(self, qty):
+        self.ensure_one()
+        res = super(SaleOrderLine, self)._prepare_invoice_line(qty)
+        if self.must_have_dates:
+            res.update({
+                'start_date': self.start_date,
+                'end_date': self.end_date,
+                })
+        return res
+
+    @api.onchange('end_date')
+    def end_date_change(self):
```

Comment:
1st This onchange methods can be merged into one
2nd Throw them away and make a number of days a computed field with and inverse method to make it editable and change start/end dates instead of a bunch of onchanges that will be a hell to test